src/storage/s3.rs

```diff
@@ -1,50 +1,77 @@
 use super::{Blob, StorageTransaction};
+use crate::Config;
 use chrono::{DateTime, NaiveDateTime, Utc};
 use failure::Error;
 use futures_util::{
     future::TryFutureExt,
     stream::{FuturesUnordered, StreamExt},
 };
 use log::warn;
-use once_cell::sync::Lazy;
-use rusoto_core::region::Region;
+use rusoto_core::{region::Region, RusotoError};
 use rusoto_credential::DefaultCredentialsProvider;
-use rusoto_s3::{GetObjectRequest, PutObjectRequest, S3Client, S3};
+use rusoto_s3::{
+    DeleteObjectsRequest, GetObjectError, GetObjectRequest, ListObjectsV2Request, ObjectIdentifier,
+    PutObjectRequest, S3Client, S3,
+};
 use std::{convert::TryInto, io::Write};
 use tokio::runtime::Runtime;
 
-#[cfg(test)]
-mod test;
-#[cfg(test)]
-pub(crate) use test::TestS3;
-
-pub(crate) static S3_BUCKET_NAME: &str = "rust-docs-rs";
-pub(crate) static S3_RUNTIME: Lazy<Runtime> =
-    Lazy::new(|| Runtime::new().expect("Failed to create S3 runtime"));
-
-pub(crate) struct S3Backend {
-    pub client: S3Client,
+pub(super) struct S3Backend {
+    client: S3Client,
+    runtime: Runtime,
     bucket: String,
+    #[cfg(test)]
+    temporary: bool,
 }
 
 impl S3Backend {
-    pub(crate) fn new(client: S3Client, bucket: &str) -> Self {
-        Self {
-            client,
-            bucket: bucket.into(),
+    pub(super) fn new(client: S3Client, config: &Config) -> Result<Self, Error> {
+        let runtime = Runtime::new()?;
+
+        // Create the temporary S3 bucket during tests.
+        if config.s3_bucket_is_temporary {
+            if cfg!(not(test)) {
+                panic!("safeguard to prevent creating temporary buckets outside of tests");
+            }
```

Comment:
Can we instead enforce that this won't compile unless it's in test mode? Something like this:

```suggestion
        extern crate criterion;
```

That should give a compile error without dev-dependencies even if the `cfg(test)` somehow disappears.

Reply:
We could also use a compile error within the function
```suggestion
            #[cfg(not(test))]
            compile_error!("You fool, you'll kill us all!");
```

Reply:
It's already `#[cfg(test)]`, I don't see a reason to add even that.